nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,535 @@
+
+  // Destroying bond connection
+  destroyBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionMonitorNode::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+
+  tf_listener_.reset();
+  tf_buffer_.reset();
+
+  odom_sub_.reset();
+  cmd_vel_in_sub_.reset();
+
+  polygon_pub_.reset();
+  cmd_vel_out_pub_.reset();
+
+  polygon_pub_timer_.reset();
+
+  polygons_.clear();
+  sources_.clear();
+
+  resetVelocity();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionMonitorNode::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+void CollisionMonitorNode::publishPolygons()
+{
+  if (!polygon_topic_.length()) {
+    // Topic is not specified: there is no need to publish
+    return;
+  }
+
+  // Fill PolygonStamped struct
+  std::unique_ptr<geometry_msgs::msg::PolygonStamped> poly_s =
+    std::make_unique<geometry_msgs::msg::PolygonStamped>();
+  poly_s->header.stamp = this->now();
+  poly_s->header.frame_id = base_frame_id_;
+  poly_s->polygon = polygons_pub_;
+
+  // Publish polygon
+  polygon_pub_->publish(std::move(poly_s));
+}
+
+void CollisionMonitorNode::odomCallback(nav_msgs::msg::Odometry::ConstSharedPtr msg)
+{
+  std::lock_guard<mutex_t> lock(velocity_mutex_);
+
+  double dt;
+  try {
+    dt = (this->now() - msg->header.stamp).seconds();
+  } catch (const std::exception & ex) {
+    RCLCPP_WARN(
+      get_logger(),
+      "Can not get difference between collision monitor and odom time: %s",
+      ex.what());
+    dt = 0.0;
+  }
+
+  if (std::fabs(dt) <= max_time_shift_) {
+    velocity_.x = msg->twist.twist.linear.x;
+    velocity_.y = msg->twist.twist.linear.y;
+    velocity_.tw = msg->twist.twist.angular.z;
+    velocity_valid_ = true;
+  } else {
+    // Warn and ignore the odom data if it and Collision Monitor node have different time stamps
+    RCLCPP_WARN(
+      get_logger(),
+      "Odom and collision monitor node timestamps differ on %f seconds. Ignoring odom.",
+      dt);
+    velocity_valid_ = false;
+  }
+}
+
+Velocity CollisionMonitorNode::getVelocity()
+{
+  std::lock_guard<mutex_t> lock(velocity_mutex_);
+  return velocity_;
+}
+
+bool CollisionMonitorNode::velocityValid()
+{
+  std::lock_guard<mutex_t> lock(velocity_mutex_);
+  return velocity_valid_;
+}
+
+void CollisionMonitorNode::resetVelocity()
+{
+  std::lock_guard<mutex_t> lock(velocity_mutex_);
+  velocity_ = {0.0, 0.0, 0.0};
+  velocity_valid_ = false;
+}
+
+void CollisionMonitorNode::cmdVelInCallback(geometry_msgs::msg::Twist::ConstSharedPtr msg)
+{
+  std::lock_guard<mutex_t> lock(cmd_vel_in_mutex_);
+  cmd_vel_in_ = {msg->linear.x, msg->linear.y, msg->angular.z};
+
+  workerMain();
+}
+
+Velocity CollisionMonitorNode::getCmdVelIn()
+{
+  std::lock_guard<mutex_t> lock(cmd_vel_in_mutex_);
```

Comment:
There's lots of these functions where these are used in only 1 place in the class itself where it can just access `cmd_vel_in_` directly, if I'm right that the mutexes are not required

Reply:
Done